File: backend/routes/studentRoutes.js
const express = require('express');
const router = express.Router();
const Quiz = require('../models/Quiz');
const QuizAttempt = require('../models/QuizAttempt');

// Helper: check if quiz is accessible
const getActiveQuiz = async (shareLink) => {
  const quiz = await Quiz.findOne({ shareLink });
  if (!quiz) return { error: 'Quiz not found', status: 404 };
  if (quiz.status === 'expired') return { error: 'This quiz has expired', status: 410 };
  if (quiz.settings.expiresAt && new Date() > quiz.settings.expiresAt) {
    quiz.status = 'expired';
    await quiz.save();
    return { error: 'This quiz has expired', status: 410 };
  }
  return { quiz };
};

// GET /api/student/quiz/:shareLink — Fetch quiz (no correct answers)
router.get('/quiz/:shareLink', async (req, res) => {
  const { quiz, error, status } = await getActiveQuiz(req.params.shareLink);
  if (error) return res.status(status).json({ message: error });

  // Strip correct answers and explanations from response
  const safeQuestions = quiz.questions.map((q, i) => ({
    index: i,
    question: q.question,
    options: q.options,
    type: q.type,
  }));

  res.json({
    _id: quiz._id,
    title: quiz.title,
    settings: {
      timeLimit: quiz.settings.timeLimit,
      questionCount: safeQuestions.length,
      tabSwitchLimit: quiz.settings.tabSwitchLimit,
    },
    questions: safeQuestions,
  });
});

// POST /api/student/quiz/:shareLink/check — Pre-attempt gate (check if email already attempted)
router.post('/quiz/:shareLink/check', async (req, res) => {
  const { email } = req.body;
  if (!email) return res.status(400).json({ message: 'Email is required' });

  const { quiz, error, status } = await getActiveQuiz(req.params.shareLink);
  if (error) return res.status(status).json({ message: error });

  const existing = await QuizAttempt.findOne({
    quizId: quiz._id,
    studentEmail: email.toLowerCase().trim(),
  });

  res.json({ attempted: !!existing });
});

// POST /api/student/quiz/:shareLink/submit — Submit quiz answers
router.post('/quiz/:shareLink/submit', async (req, res) => {
  const { studentName, studentEmail, answers, tabSwitchCount = 0, autoSubmitted = false } = req.body;

  if (!studentName || !studentEmail) {
    return res.status(400).json({ message: 'Name and email are required' });
  }
  if (!answers || !Array.isArray(answers)) {
    return res.status(400).json({ message: 'Answers array is required' });
  }

  const { quiz, error, status } = await getActiveQuiz(req.params.shareLink);
  if (error) return res.status(status).json({ message: error });

  // Double-guard: check if already attempted
  const existing = await QuizAttempt.findOne({
    quizId: quiz._id,
    studentEmail: studentEmail.toLowerCase().trim(),
  });
  if (existing) {
    return res.status(403).json({ message: 'You have already attempted this quiz' });
  }

  // Evaluate answers
  let score = 0;
  const evaluated = quiz.questions.map((q, i) => {
    const submitted = answers[i]?.selected || null;
    const isCorrect = submitted === q.correct;
    if (isCorrect) score++;
    return {
      questionIndex: i,
      selected: submitted,
      correct: q.correct,
      isCorrect,
      question: q.question,
      options: q.options,
      explanation: q.explanation,
      type: q.type,
    };
  });

  // Store attempt
  const attempt = await QuizAttempt.create({
    quizId: quiz._id,
    studentName: studentName.trim(),
    studentEmail: studentEmail.toLowerCase().trim(),
    answers: answers.map((a, i) => ({ questionIndex: i, selected: a?.selected || null })),
    score,
    totalQuestions: quiz.questions.length,
    tabSwitchCount: Number(tabSwitchCount),
    autoSubmitted,
  });

  // Build result based on visibility settings
  const result = {
    attemptId: attempt._id,
    totalQuestions: quiz.questions.length,
    tabSwitchCount: attempt.tabSwitchCount,
    autoSubmitted: attempt.autoSubmitted,
  };

  if (quiz.settings.showScore) {
    result.score = score;
    result.percentage = Math.round((score / quiz.questions.length) * 100);
  }

  if (quiz.settings.showAnswers) {
    result.answers = evaluated.map((e) => ({
      questionIndex: e.questionIndex,
      question: e.question,
      options: e.options,
      selected: e.selected,
      correct: e.correct,
      isCorrect: e.isCorrect,
      type: e.type,
      explanation: quiz.settings.showExplanations ? e.explanation : undefined,
    }));
  }

  result.visibility = {
    showScore: quiz.settings.showScore,
    showAnswers: quiz.settings.showAnswers,
    showExplanations: quiz.settings.showExplanations,
  };

  res.status(201).json({ message: 'Quiz submitted successfully', result });
});

module.exports = router;

File: backend/routes/student_routes.py
from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from models.quiz import Quiz, QuizAttempt, AnswerRecord
from pymongo.errors import DuplicateKeyError

router = APIRouter(prefix="/api/student", tags=["student"])


# ── Helpers ───────────────────────────────────────────────────────────────────

async def get_active_quiz(share_link: str) -> Quiz:
    """Fetch quiz by share link and check expiry."""
    quiz = await Quiz.find_one(Quiz.share_link == share_link)
    if not quiz:
        raise HTTPException(status_code=404, detail="Quiz not found")
    if quiz.status == "expired":
        raise HTTPException(status_code=410, detail="This quiz has expired")
    if quiz.settings.expires_at and datetime.utcnow() > quiz.settings.expires_at:
        quiz.status = "expired"
        await quiz.save()
        raise HTTPException(status_code=410, detail="This quiz has expired")
    return quiz


# ── Pydantic schemas ──────────────────────────────────────────────────────────

class CheckRequest(BaseModel):
    email: str


class AnswerSubmit(BaseModel):
    selected: Optional[str] = None


class SubmitRequest(BaseModel):
    studentName: str
    studentEmail: str
    answers: List[AnswerSubmit]
    tabSwitchCount: int = 0
    autoSubmitted: bool = False


# ── Routes ────────────────────────────────────────────────────────────────────

@router.get("/quiz/{share_link}")
async def get_quiz_for_student(share_link: str):
    """Return quiz without correct answers or explanations."""
    quiz = await get_active_quiz(share_link)

    safe_questions = [
        {
            "index": i,
            "question": q.question,
            "options": q.options,
            "type": q.type,
        }
        for i, q in enumerate(quiz.questions)
    ]

    return {
        "_id": str(quiz.id),
        "title": quiz.title,
        "settings": {
            "timeLimit": quiz.settings.time_limit,
            "questionCount": len(safe_questions),
            "tabSwitchLimit": quiz.settings.tab_switch_limit,
        },
        "questions": safe_questions,
    }


@router.post("/quiz/{share_link}/check")
async def check_attempt(share_link: str, body: CheckRequest):
    """Pre-attempt gate — check if this email has already attempted."""
    if not body.email:
        raise HTTPException(status_code=400, detail="Email is required")

    quiz = await get_active_quiz(share_link)

    existing = await QuizAttempt.find_one(
        QuizAttempt.quiz_id == quiz.id,
        QuizAttempt.student_email == body.email.lower().strip(),
    )

    return {"attempted": existing is not None}


@router.post("/quiz/{share_link}/submit", status_code=status.HTTP_201_CREATED)
async def submit_quiz(share_link: str, body: SubmitRequest):
    """Submit quiz answers, evaluate, and store attempt."""
    if not body.studentName or not body.studentEmail:
        raise HTTPException(status_code=400, detail="Name and email are required")
    if not body.answers:
        raise HTTPException(status_code=400, detail="Answers array is required")

    quiz = await get_active_quiz(share_link)

    # Check for duplicate attempt
    existing = await QuizAttempt.find_one(
        QuizAttempt.quiz_id == quiz.id,
        QuizAttempt.student_email == body.studentEmail.lower().strip(),
    )
    if existing:
        raise HTTPException(status_code=403, detail="You have already attempted this quiz")

    # Evaluate answers
    score = 0
    evaluated = []
    for i, q in enumerate(quiz.questions):
        submitted = body.answers[i].selected if i < len(body.answers) else None
        is_correct = submitted == q.correct
        if is_correct:
            score += 1
        evaluated.append({
            "questionIndex": i,
            "selected": submitted,
            "correct": q.correct,
            "isCorrect": is_correct,
            "question": q.question,
            "options": q.options,
            "explanation": q.explanation,
            "type": q.type,
        })

    # Store attempt
    answer_records = [
        AnswerRecord(
            question_index=i,
            selected=(body.answers[i].selected if i < len(body.answers) else None),
        )
        for i in range(len(quiz.questions))
    ]

    attempt = QuizAttempt(
        quiz_id=quiz.id,
        student_name=body.studentName.strip(),
        student_email=body.studentEmail.lower().strip(),
        answers=answer_records,
        score=score,
        total_questions=len(quiz.questions),
        tab_switch_count=body.tabSwitchCount,
        auto_submitted=body.autoSubmitted,
    )
    try:
        await attempt.insert()
    except DuplicateKeyError:
        raise HTTPException(status_code=403, detail="You have already attempted this quiz")

    # Build result based on visibility settings
    result = {
        "attemptId": str(attempt.id),
        "totalQuestions": len(quiz.questions),
        "tabSwitchCount": attempt.tab_switch_count,
        "autoSubmitted": attempt.auto_submitted,
    }

    if quiz.settings.show_score:
        result["score"] = score
        result["percentage"] = round((score / len(quiz.questions)) * 100) if quiz.questions else 0

    if quiz.settings.show_answers:
        result["answers"] = [
            {
                "questionIndex": e["questionIndex"],
                "question": e["question"],
                "options": e["options"],
                "selected": e["selected"],
                "correct": e["correct"],
                "isCorrect": e["isCorrect"],
                "type": e["type"],
                **({"explanation": e["explanation"]} if quiz.settings.show_explanations else {}),
            }
            for e in evaluated
        ]

    result["visibility"] = {
        "showScore": quiz.settings.show_score,
        "showAnswers": quiz.settings.show_answers,
        "showExplanations": quiz.settings.show_explanations,
    }

    return {"message": "Quiz submitted successfully", "result": result}
